### PR TITLE
Full Node Conditions & Refactorings

### PR DESCRIPTION
**Pull request by traiansf**
The main purposes of this PR are
- advancing the proof of the inclusion of projections of traces from the composition of equivocators with a fixed equivocation constraint to the composition of regular nodes with a similar constraint.
- refactorings needed to advance the proof.

### CODE REVIEW

**Review by bmmoore — APPROVED** [review]
This looks good for setting up for more kinds of projections.

File: VLSM/ProjectionTraces.v [hunk 1]
```diff
@@ -22,43 +22,57 @@ Context
...
+  (item : composite_transition_item IM)
+  (i := projT1 (l item))
+  j
+  (e : j = i)
+  : vtransition_item (IM j)
+  :=
+  let lj := eq_rect_r _ (projT2 (l item)) e in
+  @Build_transition_item _ (type (IM j)) lj (input item) (destination item j) (output item).
+
+Definition composite_transition_item_projection
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/ProjectionTraces.v:37]: It is not used in this PR, but I'm using it in the subsequent.  I agree with you that the problem you mention does exist, and an unfolding is necessary, but I still find it more convenient to have it written like this.  Let's please let it in for now, and come back to it once you see how it's used in the next PRs.

File: VLSM/Equivocators/MessageProperties.v [hunk 2]
```diff
@@ -84,9 +54,9 @@ Proof.
       eqn:Hitem'
     ; inversion HtrX; subst; clear HtrX.
     + apply equivocator_transition_item_project_inv_messages in Hitem'.
-      destruct Hitem' as [_ [_ [_ [_ Ha]]]].
+      destruct Hitem' as [_ [_ [_ [_ [_ Ha]]]]].
       inversion Hjbs; subst.
-      * left. rewrite Ha. assumption.
+      * left. unfold field_selector. simpl in Ha. simpl.  rewrite Ha. assumption.
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocators/MessageProperties.v:59]: you are right. thanks!

File: VLSM/Equivocators/Composition/SimulatingFree/FullReplayTraces.v [hunk 3]
```diff
@@ -317,13 +321,18 @@ Proof.
...
-      (apply_plan equivocators_no_equivocations_vlsm is
+      ((apply_plan equivocators_no_equivocations_vlsm is
       (map
          (transition_item_to_plan_item
-            equivocators_no_equivocations_vlsm) epref))
-      as tr_is_epref.
-    destruct tr_is_epref as (epref_items, epref_final).
+            equivocators_no_equivocations_vlsm) epref))) as tr_is_epref.
+    simpl in Heqtr_is_epref, Heqtr_is_epref'.
+    match type of Heqtr_is_epref' with
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocators/Composition/SimulatingFree/FullReplayTraces.v:329]: When it comes to terms containing `map` and other list operations (e.g., `last`), remember doesn't always replace all instances of the term, and rewrite doesn't always apply. In this particular case even `replace` doesn't seem to work properly, so I'm doing a workaround.  probably not the most elegant solution, but it saves me from having to analyze the full lowlevel contents and unfolding the right definitions to make it work.  If you of @palmskog  have any suggestions here, I'd be glad to learn some more.

File: VLSM/Equivocators/Composition/LimitedEquivocation/FixedEquivocation.v [hunk 4]
```diff
@@ -216,4 +231,234 @@ Section fixed_equivocation_with_fullnode.
...
+  (Free_no_additional_equivocation_decidable := no_additional_equivocations_dec Free comopsite_initial_decidable)
+  .
+
+Existing Instance equivocators_free_Hbs.
+
+Lemma fixed_equivocators_initial_state_project
+  (es : vstate XE)
+  (Hes : vinitial_state_prop XE es)
+  (eqv_choice : equivocators_choice IM)
+  (Heqv : proper_equivocators_choice IM eqv_choice es)
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocators/Composition/LimitedEquivocation/FixedEquivocation.v:278]: Thanks. I tried to do that. Also added more documentation.

File: VLSM/Equivocators/Common.v [hunk 5]
```diff
@@ -181,6 +202,20 @@ Next Obligation.
   lia.
 Defined.
 
+Lemma equivocator_state_extend_project_last
+  (bs : equivocator_state)
+  (s : vstate X)
+  (ns := projT1 bs)
+  : let ext := equivocator_state_extend bs s in
+    forall (l : S ns < S (projT1 ext)), projT2 ext (of_nat_lt l) = s.
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocators/Common.v:210]: Yes, I think that the initial intent was to use it as a rewrite rule. However, I see that I'm not using it in subsequent PRs so I'll remove it for now

File: VLSM/Equivocation.v [hunk 6]
```diff
@@ -2048,3 +2083,140 @@ Section full_node_constraint.
...
 
 End full_node_constraint.
+
+Section seeded_composite_vlsm_no_equivocation.
+
+(** ** Pre-loading a VLSM composition with no equivocations constraint
+
+When adding initial messages to a VLSM composition with a no equivocation
+constraint, we cannot simply use the [vlsm_add_initial_messages] construct
+because the no-equivocation constraint must also be altered to reflect that
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocation.v:2093]: the main problem is that the existing `no_equivocation_constraint` already refers to the previously defined initial_messages (the default ones for the composition). Hence, when adding new initial messages we must adjust the constraint to also allow the newly added ones.

File: VLSM/Common.v [hunk 7]
```diff
@@ -2251,6 +2265,56 @@ is also available to Y.
     Qed.
 
     (** VLSM inclusion specialized to finite trace. *)
+
+    Lemma VLSM_incl_finite_protocol_trace
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Common.v:2269]: I guess that should be `VLSM_incl_finite_protocol_trace_from`

File: VLSM/Equivocators/Common.v [hunk 8]
```diff
@@ -121,6 +121,27 @@ Definition equivocator_state_project
...
+Definition equivocator_state_descriptor_project
+  (s : equivocator_state)
+  (descriptor : MachineDescriptor)
+  : vstate X
+  :=
+  match descriptor with
+  | NewMachine sn => sn
+  | Existing j _ =>
+    let (n, bs) := s in
+    match (le_lt_dec (S n) j) with
```
[the remark this answers is not in the document]
Comment (traiansf) [reply VLSM/Equivocators/Common.v:139]: I've rewrite it so it used the `equivocator_state_project` definition above. That should allow for more independent properties to be developed eventually.

**Comment by traiansf** [comment]
@bmmoore I think I've addressed most of your comments. Please take another look.

**Review by bmmoore — APPROVED** [review]
Changes look fine. You said there was more work following after this, so I wasn't expecting to change anything inessential before merging this PR anyway.

**Pull request merged**